Read a previously generated `META-INF/quarkus-build-steps.list` file …

lucamolteni opened the pull request:
Read a previously generated `META-INF/quarkus-build-steps.list` file in ExtensionBuildProcessor.java to support incremental compilation of extensions.

* This is useful while developing extensions in Intellij IDEA and developers want to run tests from the IDE

lucamolteni committed:
Read a previously generated `META-INF/quarkus-build-steps.list` file in ExtensionBuildProcessor.java to support incremental compilation of extensions.
* This is useful while developing extensions in Intellij IDEA and developers want to run tests from the IDE
* read => readSet
  write => writeSet
  write(properties) => writeProperties

 Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

FroMage reviewed (commented):
This is a great fix we've all been waiting for!

I suspect this will break when we do an initial build with a file containing say 2 build steps, and then we remove one of the build steps, because the initial list will contain the two, the incremental compilation round will detect the one remaining (which is already in the previous set) and the removed build step will not be removed, so we'll have a build step list with a removed build step.

This is just a guess, though, remains to be checked if that is actually a problem in reality or not :-/

gsmet commented on core/processor/src/main/java/io/quarkus/annotation/processor/extension/ExtensionBuildProcessor.java:
```diff
@@ -67,7 +68,23 @@ public void process(Set<? extends TypeElement> annotations, RoundEnvironment rou
+         */
+        List<String> allProcessors = utils.filer().read(Outputs.META_INF_QUARKUS_BUILD_STEPS);
+
+        Set<String> allBuildSteps = new HashSet<String>(allProcessors);
```
```suggestion
        Set<String> allBuildSteps = new TreeSet<>(allProcessors);
```
(currently working on build reproducibility :))

gsmet commented on core/processor/src/main/java/io/quarkus/annotation/processor/util/FilerUtil.java:
```diff
@@ -56,6 +61,25 @@ public void write(String filePath, String value) {
         }
     }
 
+    public List<String> read(String filePath) {
```
I fear using `read` is going to be annoying if at some point, we want to have a method to read a string.

We should also rename `void write(String filePath, Set<String> set)` to `writeSet`. And `write(String filePath, Properties properties)` to `writeProperties`.

```suggestion
    public Set<String> readSet(String filePath) {
```

gsmet commented on core/processor/src/main/java/io/quarkus/annotation/processor/util/FilerUtil.java:
```diff
@@ -56,6 +61,25 @@ public void write(String filePath, String value) {
     }
 
+    public List<String> read(String filePath) {
+        List<String> output = new ArrayList<>();
```
```suggestion
        Set<String> output = new TreeSet<>();
```

gsmet commented on core/processor/src/main/java/io/quarkus/annotation/processor/util/FilerUtil.java:
```diff
@@ -56,6 +61,25 @@ public void write(String filePath, String value) {
+            return output;
+        } catch (IOException e) {
+            processingEnv.getMessager().printMessage(Kind.WARNING, "Failed to read " + filePath + ": " + e);
+            return Collections.emptyList();
```
```suggestion
            return Set.of();
```

gsmet reviewed (commented):
Thanks!

I added some suggestions. If you don't have the time to apply them, let me know and I'll finalize this PR.

holly-cummins commented:
I have no comment on the code, but I'll just add that I'm so happy to see this fix, since it makes it possible to (a) run the tests in the Quarkus codebase in the debugger (b) not have to rerun a full build after attempting to run tests in the debugger. I've been missing that ability. Excited for this to be merged!

lucamolteni commented:
> This is a great fix we've all been waiting for!
> 
> I suspect this will break when we do an initial build with a file containing say 2 build steps, and then we remove one of the build steps, because the initial list will contain the two, the incremental compilation round will detect the one remaining (which is already in the previous set) and the removed build step will not be removed, so we'll have a build step list with a removed build step.
> 
> This is just a guess, though, remains to be checked if that is actually a problem in reality or not :-/

Yes, I assumed that deleting new Processors is less frequent than modifying one

I'm going to test it to see if I can find a workaround for that as well

FroMage commented:
> Yes, I assumed that deleting new Processors is less frequent than modifying one

s/Processor/build step/

I'm not sure it's that infrequent. I just wonder if it indeed breaks everything if we delete one (this is just removing a single build step method in a processor), and how the user can recover in such a case.

gsmet committed:
Avoid logging when the file does not exist

This is a perfectly valid use case when you do a clean or start with a
clean repo.

gsmet committed:
Avoid the confusion between build steps and processors

The filename is a bit unfortunate but it's how it is, given it has been
like that since forever and there's no value breaking compatibility.

gsmet commented:
> s/Processor/build step/

Actually, what is a bit confusing is that what we push to this build steps file are actually the name of the processors (the classes containing the build steps).
So deleting a build step is not a problem.
Deleting the whole processor might be, but it's a lot less common.

I added a commit to clarify this a bit.

I also dropped the warning when there's no existing file, given it's a valid use case.

@lucamolteni could you have a look and make sure I didn't break your use case? Thanks!

FroMage commented on core/processor/src/main/java/io/quarkus/annotation/processor/extension/ExtensionBuildProcessor.java:
```diff
@@ -67,8 +67,22 @@ public void process(Set<? extends TypeElement> annotations, RoundEnvironment rou
+         * creating a subset of processors that are not enough to run a quarkus app
+         * By assuming a full compilation was made initially, all the processors are included inside the
+         * META-INF/quarkus-build-steps.list file
+         * So by reading it we can ensure that all the build steps are included.
```
```suggestion
         * So by reading it we can ensure that all the processors are included.
```

lucamolteni commented:
> @lucamolteni could you have a look and make sure I didn't break your use case? Thanks!

Yes it's still working

lucamolteni commented:
> > Yes, I assumed that deleting new Processors is less frequent than modifying one
> 
> s/Processor/build step/
> 
> I'm not sure it's that infrequent. I just wonder if it indeed breaks everything if we delete one (this is just removing a single build step method in a processor), and how the user can recover in such a case.

I've just tried creating a new Processor (with two build steps and two build items), adding and modifying the processor worked flawlessly, removing the processor threw a `ClassNotFoundException(name);`

The fix is pretty easy, just remove the class from `quarkus-build-steps.list` or run `mvn package` , perhaps we could notify the user on how to do that? 

Otherwise we can be even more defensive and catching the Exception in io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:580

gsmet commented:
> Otherwise we can be even more defensive and catching the Exception in io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:580

Yeah this is where I draw the line :).

gsmet reviewed (approved):
Let's merge it once CI is green.

Thanks for fighting for it :).

FroMage commented:
> I've just tried creating a new Processor (with two build steps and two build items), adding and modifying the processor worked flawlessly, removing the processor threw a `ClassNotFoundException(name);`

OK, thanks for testing this.

> The fix is pretty easy, just remove the class from `quarkus-build-steps.list` or run `mvn package` , perhaps we could notify the user on how to do that?

It would be nice if the exception did include a message pointing to this indeed, because it's not going to be obvious to anyone but us three. Do you have the full strack trace by chance?

> Otherwise we can be even more defensive and catching the Exception in io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:580

Nah, agreed with @gsmet I don't think this is where it should be handled.
